crates/rpc/src/server/accept.rs

```diff
@@ -1,61 +1,186 @@
+/// constraints then the request is rejected.
+///
+/// Note that both values are optional, as is the header itself. If unset, the server considers
+/// any value acceptable.
+///
+/// As part of the accept header's standard, all media ranges are examined in quality weighting
+/// order until a matching content type is found. This means that the client can set multiple
+/// `application/vnd.miden` values and each will be tested until one passes. If none pass, the
+/// request is rejected.
+///
+/// ## Format
+///
+/// Parameters are optional and order is not important.
+///
+/// ```
+/// application/vnd.miden; version=<version-req>; genesis=0x1234
+/// ```
 #[derive(Clone)]
-pub struct AcceptLayer {
-    version_req: VersionReq,
+pub struct AcceptHeaderLayer {
+    supported_versions: VersionReq,
+    genesis_commitment: Word,
 }
 
-impl AcceptLayer {
-    /// Create a new accept layer that validates version values specified in the HTTP ACCEPT header.
-    ///
-    /// The version requirement is based on the version field found in the workspace's Cargo.toml.
-    ///
-    /// # Panics:
-    ///
-    /// Panics if the version string in Cargo.toml is not valid semver. The version string is made
-    /// into an env var at compile time which means that the unit tests prove this cannot panic
-    /// in practice.
-    pub fn new() -> anyhow::Result<Self> {
-        // Parse the full version string (e.g. "0.8.0").
-        let version = env!("CARGO_PKG_VERSION");
-        let version = Version::parse(version)?;
-
-        // Form a version requirement from the major and minor version numbers.
-        let version_req = format!("={}.{}", version.major, version.minor);
-        let version_req = VersionReq::parse(&version_req).expect("valid version requirement");
-        Ok(AcceptLayer { version_req })
+#[derive(Debug, thiserror::Error)]
+enum AcceptHeaderError {
+    #[error("header value could not be parsed as a UTF8 string")]
+    InvalidUtf8(#[source] ToStrError),
+
+    #[error("accept header's media type could not be parsed")]
+    InvalidMediaType(#[source] mediatype::MediaTypeError),
+
+    #[error("a Q value was invalid")]
+    InvalidQValue(#[source] QParsingError),
+
+    #[error("version value failed to parse")]
+    InvalidVersion(#[source] semver::Error),
+
+    #[error("genesis value failed to parse")]
+    InvalidGenesis(#[source] WordError),
+
+    #[error("server does not support any of the specified application/vnd.miden content types")]
+    NoSupportedMediaRange,
+}
+
+impl AcceptHeaderLayer {
+    pub fn new(rpc_version: &Version, genesis_commitment: Word) -> Self {
+        let supported_versions = VersionReq {
+            comparators: vec![Comparator {
+                op: semver::Op::Exact,
+                major: rpc_version.major,
+                minor: rpc_version.minor.into(),
+                patch: None,
+                pre: semver::Prerelease::default(),
+            }],
+        };
```

Comment:
I'm a bit on the fence about this; but I've left the current behavior as is.

The issue is that its a bit back-to-front. The client should be indicating which versions they're compatible with. For example, their RPC version might be patch compatible with our server's but they may require a new feature or addition that we introduced.

Its easy to do by swapping the client and server types around. The downside is that client's need to think a bit more about what they want.

Reply:
I think this is fine for now, but I would create an issue to revisit this in the future because in principle I agree: the client should be specifying the version that it is compatible with and if the node is within this specifications, the request should be accepted.